flux-jobs: small functionality and testing updates

grondo opened the pull request:
This PR is split off from work in #3045.

It includes the custom `HeaderFormatter` class for `flux-jobs`, the update to force UTF-8 output encoding, and removal of some unused code.

However, the majority of changes here is to the `t2800-jobs-cmd.t` test, which by the end was largely rewritten to get rid of hardcoded job counts and redundant calls to `flux-jobs` which made the test one of the longest in the suite.

The resulting runtime is reduced by half.

Before:
```
$ time ./t2800-jobs-cmd.t 
[snip]
real	0m52.974s
user	0m48.355s
sys	0m8.883s
```

After:
```
$ time ./t2800-jobs-cmd.t 
[snip]
real	0m28.932s
user	0m26.798s
sys	0m5.828s
```

grondo commented on t/t2800-jobs-cmd.t:
```diff
@@ -45,60 +63,95 @@ wait_states() {
+	#  Synchronize all expected states
 	wait_states
+	#
+	#  Sort inactive.ids in expected order, since we cant accurately
```
"can't" have an unbalanced single quote inside the test  ;-) I'll change it to cannot.

chu11 replied:
oh yeah, doh!

grondo commented:
It appears most of the tests are failing in Travis. I guess I have more work to do here.

grondo commented:
> This is awesome cleanup @grondo. Thanks for cleaning up all of this mess which is largely my fault, built up over many incremental changes ever since the first flux job list tests in (now) t2230-job-info-list.t.

No problem! I could tell the tests grew over time :wink:

grondo commented:
> also, if you think this PR is about ready, I can likely update t2230-job-info-list in a similar way and add it on top of this PR (or I can do it in another PR too).

Yeah, I think if I can fix the tests failing in Travis, this one will be done for now.

grondo commented:
Ok, the problem was that the data ready from psuedo-tty in `runpty.py` is fully buffered, and thus lines of output in the "asciicast" output format could be split across multiple asciicast entries, or multiple lines could appear in one entry.

Instead of figuring out how to work around this problem, I added a `--line-buffer` option to `runpty.py` that attempts to buffer "lines" of output (`"\r\n"`) before writing to the underlying formatter class. This should ensure that each line of output will appear in its own asciicast entry, thus allowing `grep` to work a little more reliably.

chu11 commented:
> Ok, the problem was that the data ready from psuedo-tty in runpty.py is fully buffered, and thus lines of output in the "asciicast" output format could be split across multiple asciicast entries, or multiple lines could appear in one entry.

Was this unique to your changes or something that was there before, but we got lucky?

chu11 commented:
i think things look good, I approved if you think its ready for a MWP.

chu11 commented on t/t2800-jobs-cmd.t:
```diff
@@ -27,83 +27,132 @@ runpty="${SHARNESS_TEST_SRCDIR}/scripts/runpty.py -f asciicast"
+	     || [ "$(flux jobs -n --filter=inactive | wc -l)" != "$inactive" ]) \
+	&& [ $i -lt 50 ]
+	do
+	sleep 0.1
```
doh!  I just noticed these indentations looked off

grondo replied:
Ok, thanks! I'll fix that up.

grondo commented:
Ok, @chu11. Fixed some indent issues including the one you caught (rogue search/replace seemed to have "flattened" multiple non-tab indents into a single tab indent)

Squashed the result.

chu11 commented:
> Edit: Oh wait I forgot, I haven't started fixing up t2230-job-info-list yet. Do you want to merge so we can all move forward with our PRs and just setup an issue for t2230-job-info-list for later?

As an side, looking at `t2230-job-info-list` the gains to changing the tests there are much smaller than for `t2800-jobs-cmd`.  But some minor cleanups can happen.  I'll go ahead and add it on top of my #3062 PR.

grondo commented:
Hm, travis doesn't seem to want to start testing this PR. I'll kick it.

chu11 commented on t/t2800-jobs-cmd.t:
```diff
@@ -27,83 +27,131 @@ runpty="${SHARNESS_TEST_SRCDIR}/scripts/runpty.py -f asciicast"
+	flux job wait-event $jobid clean &&
+	echo $jobid >> inactive.ids &&
+	echo $jobid > cancelled.ids &&
+	cat inactive.ids active.ids >> all.ids &&
```
as I'm cleaning up `t2230-job-info-list`, noticed that you don't use `all.ids` really anywhere.  And `all.ids` doesn't seem to contain pending or running ids.

Perhaps construct an `all.ids` that can be used in the future?  I'm doing

```
        #                                                                                                                                   
        #  Create an all.ids list                                                                                                           
        #                                                                                                                                   
        cat pending.ids > all.ids &&                                                                                                        
        cat running.ids >> all.ids &&                                                                                                       
        cat inactive.ids >> all.ids &&  
```

chu11 replied:
nevermind, I misread your line

grondo replied:
This went through many many iterations, so I would not be surprised if there are some unused things or inconsistencies.

grondo commented:
(sorry I had this comment pending and never hit the comment button)

> Was this unique to your changes or something that was there before, but we got lucky?

I think it was unique to the new tests. I collapsed several `flux jobs` runs into one with `-a` so there were many lines that needed to match. I think the previous tests just took a sample of the asciicast output as well.

grondo commented:
Ok, I can't get travis to start running tests. Maybe I'll try pushing an inane change and see if that kicks it.

grondo commented:
Pushed (and squashed) a couple chain-lint fixes. Travis-CI still refuses to build this PR. We'll see if resolves itself overnight I guess.

chu11 commented:
@grondo i ended up finishing up my cleanup of `t2230-job-info-list` last night / this morning.  I could add on top of this PR?  Dunno if I can push to your repo or not,.  So I just a pull request on your repo in your tree: https://github.com/grondo/flux-core/pull/15

grondo commented:
You should be able to push directly to my branch. However, I've cherry-picked your commit and will push it here.

grondo commented:
Hm, got two different test failures in unrelated tests. Restarted the  builds to see if the failures were spurious.

grondo commented:
@chu11, if tests pass I think this PR is ready, so I was going to set MWP. Ok with you?

chu11 commented:
> @chu11, if tests pass I think this PR is ready, so I was going to set MWP. Ok with you?

Yup.  Lets get this in!
